clash-lib/src/Clash/Backend/VHDL.hs

```diff
@@ -1508,12 +1557,24 @@ expr_ _ (DataCon ty@(Vector 1 elTy) _ [e])       = do
 expr_ _ e@(DataCon ty@(Vector _ elTy) _ [e1,e2]) = do
   syn <- Ap hdlSyn
   case syn of
+    -- When targeting Vivado, arrays must use std_logic_vector for elements.
     Vivado -> qualTyName ty <> "'" <> case vectorChain e of
-      Just es -> align (tupled (mapM (toSLV elTy) es))
+      -- Rendering a sum type results in ambiguity here, as the elements of the
+      -- array are rendered as
+      --
+      --   toSLV(fromSLV(x))
+      --
+      -- This can be avoided by checking if the expression is wrapped in FromBv
+      -- and if so dropping the fromBv and not inserting a toSLV call.
+      --
+      Just es -> align (tupled (mapM slvArrayElem es))
```

Comment:
What about doing the following instead:
```diff
diff --git a/clash-lib/src/Clash/Backend/VHDL.hs b/clash-lib/src/Clash/Backend/VHDL.hs
index 9cce9490..79a15c13 100644
--- a/clash-lib/src/Clash/Backend/VHDL.hs
+++ b/clash-lib/src/Clash/Backend/VHDL.hs
@@ -1567,14 +1567,11 @@ expr_ _ e@(DataCon ty@(Vector _ elTy) _ [e1,e2]) = do
       -- This can be avoided by checking if the expression is wrapped in FromBv
       -- and if so dropping the fromBv and not inserting a toSLV call.
       --
-      Just es -> align (tupled (mapM slvArrayElem es))
+      Just es -> align (tupled (mapM (toSLV elTy) es))
       Nothing -> parens ("std_logic_vector'" <> parens (toSLV elTy e1) <+> "&" <+> expr_ False e2)
     _ -> qualTyName ty <> "'" <> case vectorChain e of
             Just es -> align (tupled (mapM (expr_ False) es))
             Nothing -> parens (qualTyName elTy <> "'" <> parens (expr_ False e1) <+> "&" <+> expr_ False e2)
- where
-  slvArrayElem (FromBv _ _ x) = expr_ False x
-  slvArrayElem x = toSLV elTy x
 
 expr_ _ (DataCon ty@(RTree 0 elTy) _ [e]) = do
   syn <- Ap hdlSyn
@@ -1719,10 +1716,12 @@ expr_ _ (ToBv topM hwty e) = do
                parens (qualTyName hwty <> "'" <> parens (expr_ False e))
     Just t  -> pretty t <> dot <> pretty t <> "_types" <> dot <> "toSLV" <> parens (expr_ False e)
 
-expr_ _ (FromBv topM _ e) = do
+expr_ _ (FromBv topM hwty e) = do
   nm <- Ap $ use modNm
-  maybe (pretty nm <> "_types" ) (\t -> pretty t <> dot <> pretty t <> "_types") topM <> dot <>
-    "fromSLV" <> parens (expr_ False e)
+  qualTyName hwty <> "'" <> parens
+    (maybe (pretty nm <> "_types")
+           (\t -> pretty t <> dot <> pretty t <> "_types") topM <> dot <>
+           "fromSLV" <> parens (expr_ False e))
 
 expr_ _ e = error $ $(curLoc) ++ (show e) -- empty
 ```
That seems less adhoc and prevents the ambiguity in all situations where `FromBv` is used.

Reply:
That being said, we could actually keep your "adhoc" solution as an optimization.

Reply:
This is what I get for not having GitHub and my terminal visible at once :cry: 
Would you like me to change it back?

Reply:
No, it's fine.